### PR TITLE
Don't allow fixes to COPY code from templated regions

### LINKED ISSUE #2051: sqlfluff fix with dbt templater replaces aliases with sources
## Expected Behaviour

No replacement of aliases with the source, because the source database and schema changes between environments.

## Observed Behaviour
```
SELECT t.dv_RepositoryID
FROM {{ source('mydb', 'repo') }} t
```
is replaced with
```
SELECT mydb_dev.repo.dv_repositoryid
FROM {{ source('mydb', 'repo') }} t
```

However `mydb_dev` is the development database.

## Dialect
snowflake

## Version
sqlfluff, version 0.8.2
python 3.9

## Configuration

```
[tool.sqlfluff.core]
templater = "dbt"
sql_file_exts = ".sql,.sql.j2,.dml,.ddl"
dialect = "snowflake"
ignore_templated_areas = true

[tool.sqlfluff.indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[tool.sqlfluff.templater.dbt]
project_dir = "."
profile = "snowflake"
```

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made

Fixes #2051 

This PR builds on yesterday's PR #2220:
* Yesterday's PR discards all fixes from a `LintResult` if any of the fixes **modify** templated code.
* This PR discards all fixes from a `LintResult` if any of the fixes **read from** templated code. This required a slight change to any rules which do this. (Fortunately, there are only a handful of such rules.)

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/testing/rules.py [hunk 1]
```diff
@@ -85,12 +85,8 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
                     line_numbers, actual_line_numbers
                 )
             )
-    # The query should already have been fixed if possible so just return the raw.
-    if linted.num_violations(fixable=True) > 0:
-        fixed, _ = linted.fix_string()
-        return fixed
-    else:
-        return linted.tree.raw
+    fixed, _ = linted.fix_string()
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:88]: This code was overly complex and wrong. For tests using Jinja, we need to call `fix_string()` to get the pre-rendered , fixed code.

File: src/sqlfluff/core/rules/base.py [hunk 2]
```diff
@@ -114,11 +114,14 @@ class LintFix:
...
             it represents the segment to delete, for creations it implies the
             position to create at (with the existing element at this position
-            to be moved *after* the edit), for an `edit` it implies the segment
-            to be replaced.
-        edit (:obj:`BaseSegment`, optional): For `edit` and `create` fixes, this
-            hold the segment, or iterable of segments to create or replace at the
+            to be moved *after* the edit), for a `replace` it implies the
+            segment to be replaced.
+        edit (:obj:`BaseSegment`, optional): For `replace` and `create` fixes,
+            this holsd the iterable of segments to create or replace at the
```
Comment (tunetheweb) [inline src/sqlfluff/core/rules/base.py:120]:
```suggestion
            this holds the iterable of segments to create or replace at the
```

File: src/sqlfluff/core/rules/base.py [hunk 3]
```diff
@@ -114,11 +114,14 @@ class LintFix:
             the *position* that this fix should be applied at. For deletions
             it represents the segment to delete, for creations it implies the
             position to create at (with the existing element at this position
-            to be moved *after* the edit), for an `edit` it implies the segment
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/base.py:117]: Fix some docs that we overlooked updating after a recent API change.

File: src/sqlfluff/rules/L045.py [hunk 4]
```diff
@@ -3,10 +3,8 @@
 
 from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
 from sqlfluff.core.rules.analysis.select_crawler import Query, SelectCrawler
-from sqlfluff.core.rules.doc_decorators import document_fix_compatible
 
 
-@document_fix_compatible
```
Comment (barrywhart) [inline src/sqlfluff/rules/L045.py:9]: Unrelated fix: This rule does not generate fixes, thus the docs were wrong.

Comment (tunetheweb) [inline src/sqlfluff/rules/L045.py:9]: Why have you stopped making this fix compatible?

Comment (barrywhart) [reply src/sqlfluff/rules/L045.py:9]: It never was "fix compatible" -- if you look at the code, it never returns any `LintFix` objects. So this was always wrong. (It's not related to the rest of the PR, just a thing that I noticed when working on it.) 😛

File: src/sqlfluff/core/rules/base.py (hunk 2 above)
Comment (barrywhart) [reply src/sqlfluff/core/rules/base.py]: Good catch -- fixed!

**Review by tunetheweb — APPROVED** [review]

**Review by jpy-git — APPROVED** [review]
Looks good to me 🥇
